File: app/src/main/java/de/xikolo/storages/NotificationStorage.java
package de.xikolo.storages;

import android.content.Context;
import android.content.SharedPreferences;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

import de.xikolo.storages.base.BaseStorage;

public class NotificationStorage extends BaseStorage {

    private static final String PREF_NOTIFICATIONS = "pref_notifications";

    private String DOWNLOAD_NOTIFICATIONS = "download_notifications";

    public NotificationStorage() {
        super(PREF_NOTIFICATIONS, Context.MODE_PRIVATE);
    }

    public List<String> getDownloadNotifications() {
        Gson gson = new Gson();
        String json = preferences.getString(DOWNLOAD_NOTIFICATIONS, null);
        Type type = new TypeToken<ArrayList<String>>() {}.getType();
        return gson.fromJson(json, type);
    }

    public void saveDownloadNotifications(List<String> notifications) {
        SharedPreferences.Editor editor = preferences.edit();
        Gson gson = new Gson();
        editor.putString(DOWNLOAD_NOTIFICATIONS, gson.toJson(notifications));
        editor.commit();
    }

    public void addDownloadNotification(String notification) {
        List<String> notifications = getDownloadNotifications();

        if (notifications == null) {
            notifications = new ArrayList<>();
        }
        notifications.add(notification);

        SharedPreferences.Editor editor = preferences.edit();
        Gson gson = new Gson();
        editor.putString(DOWNLOAD_NOTIFICATIONS, gson.toJson(notifications));
        editor.commit();
    }

    public void deleteDownloadNotification(String notification) {
        List<String> notifications = getDownloadNotifications();

        if (notifications != null) {
            notifications.remove(notification);
            SharedPreferences.Editor editor = preferences.edit();
            Gson gson = new Gson();
            editor.putString(DOWNLOAD_NOTIFICATIONS, gson.toJson(notifications));
            editor.commit();
        }
    }

    public void deleteAllDownloadNotifications() {
        SharedPreferences.Editor editor = preferences.edit();
        editor.clear();
        editor.commit();
    }

}

File: app/src/main/java/de/xikolo/storages/NotificationStorage.kt
package de.xikolo.storages

import android.content.Context
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import de.xikolo.storages.base.BaseStorage
import java.util.*

class NotificationStorage : BaseStorage(PREF_NOTIFICATIONS, Context.MODE_PRIVATE) {

    val downloadNotifications: MutableList<String>
        get() {
            val json = getString(DOWNLOAD_NOTIFICATIONS)
            val type = object : TypeToken<ArrayList<String>>() {}.type
            return Gson().fromJson(json, type)
        }

    fun addDownloadNotification(notification: String) {
        var notifications: MutableList<String>? = downloadNotifications

        if (notifications == null) {
            notifications = ArrayList()
        }
        notifications.add(notification)

        putString(DOWNLOAD_NOTIFICATIONS, Gson().toJson(notifications))
    }

    fun deleteDownloadNotification(notification: String) {
        val notifications = downloadNotifications
        notifications.remove(notification)
        putString(DOWNLOAD_NOTIFICATIONS, Gson().toJson(notifications))
    }

    companion object {
        private const val PREF_NOTIFICATIONS = "pref_notifications"
        private const val DOWNLOAD_NOTIFICATIONS = "download_notifications"
    }

}
